CBOR Writer: Use canonical NaN representation for NaN values

tomeksowi opened the pull request:
[RFC 7049 (CBOR)](https://www.rfc-editor.org/rfc/rfc7049) specifies "If NaN is an allowed value, it must always be represented as 0xf97e00". The only exception I made is when the user explicitly requests precision (FP size) is preserved.

The problem occurred for x86, C# defines NaN as 0.0/0.0 which yields -NaN on x86 FP units, which gets encoded as 0xf9fe00.

Fixes issue #92080

Also, use positive NaNs on RISC-V for FP ops crossgen-evaluated when host is x86.

Part of #84834
cc @wscho77 @HJLeee @JongHeonChoi @t-mustafin @alpencolt @gbalykov @clamp03 @sirntar @yurai007

tomeksowi committed:
Use canonical NaN representation for NaN values

RFC 7049 (CBOR) specifies "If NaN is an allowed value, it must always be represented as 0xf97e00". The only exception is when the user explicitly requests precision (FP size) is preserved.

The problem occurred for x86, C# defines NaN as 0.0/0.0 which yields -NaN on x86 FP units, which gets encoded as 0xf9fe00.

Fixes issue #92080

vcsjones commented on src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.cs:
```diff
@@ -18,6 +18,11 @@ public partial class CborWriter
         /// <para>The written data is not accepted under the current conformance mode.</para></exception>
         public void WriteSingle(float value)
         {
+            if (float.IsNaN(value))
```
I see that `WriteSingle` and `WriteDouble` are covered, but not `WriteHalf`

https://github.com/dotnet/runtime/blob/88f25474334bfb0da10e0b5acb0219a64a04828b/src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.netcoreapp.cs#L20-L27

I think the appropriate change is to leave WriteSingle and WriteDouble alone, and fix WriteHalf to say "If this is NaN, then write it with positive NaN"

tomeksowi replied:
> I think the appropriate change is to leave WriteSingle and WriteDouble alone, and fix WriteHalf to say "If this is NaN, then write it with positive NaN"

It wouldn't cover writing a float/double NaN and requiring preserving precision (they should be positive too as per RFC 7049).

You're right about WriteHalf, though.

tomeksowi replied:
I pushed the positive NaN reassignment to Write(Single|Double)Core because most of the cases are handled by WriteHalf, that way code for Write(Double|Single) is left alone. Besides, it's more of an representation thing so it's probably better to handle it just before writing.

tomeksowi committed:
Use canonical CBOR (positive) NaN in WriteHalf

krwq commented on src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.cs:
```diff
@@ -57,6 +57,10 @@ public void WriteDouble(double value)
         {
+            if (float.IsNaN(value))
+            {
+                value = CborHelpers.Int32BitsToSingle(0x7fc00000); // canonical NaN as per RFC 7049
```
could these use constants? CanonicalFloat defined here or just float.NaN or something? `CborHelpers.Int32BitsToSingle(0x7fc00000);` and `BitConverter.Int64BitsToDouble(0x7ff8000000000000)` seem constant or am I missing something?

tomeksowi replied:
Sure, I put them in named constants.

`float.NaN` is defined as `0.0f / 0.0f` which doesn't yield a consistent bit pattern across platforms (it's negative on x86, positive elsewhere).

eiriktsarpalis commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -106,7 +106,7 @@ struct DoubleTraits
 template <typename TFp, typename TFpTraits>
 TFp FpAdd(TFp value1, TFp value2)
 {
-#if defined(TARGET_ARMARCH) || defined(TARGET_LOONGARCH64)
```
cc @dotnet/jit-contrib

tannergooding replied:
Why is any change to valuenum.cpp needed here for the cbor fix?

This seems like an independent change related to constant folding which should be in it's own PR.

tomeksowi replied:
OK, #93098

tomeksowi committed:
Put canonical CBOR NaNs into named constants

jakobbotsch reviewed (approved):
The JIT change LGTM.

eiriktsarpalis commented on src/libraries/System.Formats.Cbor/tests/Writer/CborWriterTests.Simple.netcoreapp.cs:
```diff
@@ -18,7 +18,7 @@ public partial class CborWriterTests
         [InlineData(0.00006103515625, "f90400")]
         [InlineData(-4.0, "f9c400")]
         [InlineData(float.PositiveInfinity, "f97c00")]
-        [InlineData(float.NaN, "f9fe00")]
```
Could you also add a few reader tests validating that the previously used encoding is still readable as NaN?

tomeksowi committed:
Add CborReader tests to verify the previously emitted negative NaN bit patterns are still readable as NaN.

tannergooding commented on src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.cs:
```diff
@@ -7,6 +7,10 @@ namespace System.Formats.Cbor
     public partial class CborWriter
     {
+        // Canonical NaN representations as per RFC 7049
+        private static readonly float CanonicalNaNSingle = CborHelpers.Int32BitsToSingle(0x7fc00000);
+        private static readonly double CanonicalNaNDouble = BitConverter.Int64BitsToDouble(0x7ff8000000000000);
```
nit: I think it would be better to simply calls these what they actually are, which is `PositiveQNaN`.

The concern being that they are canonical in the context of CBOR, but not more generally, and someone may get the wrong idea especially if just generally searching for things related to NaN in the repo.

Calling it what it is, a `+QNaN` and then leaving a comment that it's used because `RFC 7049` expects it as it's own canonical representation avoids the ambiguity, while potentially providing additional clarity and allowing us to clearly indicate why we don't use `float.NaN`

tomeksowi replied:
OK

tannergooding commented on src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.cs:
```diff
@@ -57,6 +61,10 @@ public void WriteDouble(double value)
         }
         private void WriteSingleCore(float value)
         {
+            if (float.IsNaN(value))
+            {
+                value = CanonicalNaNSingle;
+            }
```
This isn't guaranteed to work across all possible runtimes/compilers/future versions. The IEEE 754 spec allows for a `NaN` to be normalized to any representation (and while we aren't likely to do that ourselves, it is very much legal to do so).

Rather, instead, you should have something like:
```csharp
if (float.IsNaN(value))
{
    WriteInt32Core(PositiveQNaNBits);
    return;
}

...
```

tomeksowi replied:
WriteInt32Core doesn't exist but I agree, it shouldn't go through an FPU.

tannergooding commented on src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.cs:
```diff
@@ -130,7 +142,7 @@ public void WriteSimpleValue(CborSimpleValue value)
         {
             result = (float)value;
-            return BitConverter.DoubleToInt64Bits(result) == BitConverter.DoubleToInt64Bits(value);
+            return double.IsNaN(value) || BitConverter.DoubleToInt64Bits(result) == BitConverter.DoubleToInt64Bits(value);
```
Is this simply trying to see if a given `double` can be represented as a `float` without loss in precision?

Is there a non-canonicalization mode where the `NaN` payload bits are necessary to consider?

tomeksowi replied:
> Is this simply trying to see if a given `double` can be represented as a `float` without loss in precision?

That's what the hitherto bit pattern comparison was trying to do. That failed for FPUs with lax NaN propagation rules (e.g. RISC-V) because the conversion to float drops the sign and payload.

> Is there a non-canonicalization mode where the `NaN` payload bits are necessary to consider?

[RFC 8949](https://www.rfc-editor.org/rfc/rfc8949) offers an easy way out by saying "we don't support payloads":
*If NaN is an allowed value, and there is no intent to support NaN payloads or signaling NaNs, the protocol needs to pick a single representation, typically 0xf97e00. If that simple choice is not possible, specific attention will be needed for NaN handling.*

However, CborConformanceMode.Ctap2CanonicalMode "that simple choice is not possible" as it says: *representations of any floating-point values are not changed*. For that mode we don't enter TryConvertDoubleToSingle and friends so no problem there. But I guess in Write(Single|Double)Core and WriteHalf the condition should be:
```cs
if (float.IsNaN(value) && !CborConformanceModeHelpers.RequiresPreservingFloatPrecision(ConformanceMode))
{
    value = CanonicalNaNSingle;
}
```
Agreed?

tomeksowi committed:
Use only half-float canonical CBOR representation for NaNs

NaNs only get written as 4 or 8 bytes only in CTAP2 mode, which requires to preserve all bits anyway.

+ review fixes

eiriktsarpalis commented on src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.netcoreapp.cs:
```diff
@@ -21,7 +21,14 @@ public void WriteHalf(Half value)
             EnsureWriteCapacity(1 + sizeof(short));
             WriteInitialByte(new CborInitialByte(CborMajorType.Simple, CborAdditionalInfo.Additional16BitData));
-            BinaryPrimitives.WriteHalfBigEndian(_buffer.AsSpan(_offset), value);
+            if (Half.IsNaN(value) && !CborConformanceModeHelpers.RequiresPreservingFloatPrecision(ConformanceMode))
```
I'm not sure I understand why the second check is necessary. It is meant to check if the current conformance mode permits using lower-precision representation, however `Half` is the lowest one available per the spec.

tomeksowi replied:
Half NaN may also have a sign and payload and CTAP2 mode description says *representations of any floating-point values are not changed*. So I opted for writing the input verbatim.
